File: es6/lib/state.js
import { inspect } from 'util';
import Pair from './pair';
import Variable from './variable';

export default class State {
  constructor(variables = [], values = new Map()) {
    this._variables = variables;
    this._values    = values;
  }

  createVars(...names) {
    let vars = Variable.named(...names);
    return [new State(this._variables.concat(vars), this._values), vars];
  }

  assign(...values) {
    var newMap = new Map(Array.from(this._values));
    for (let [key, value] of values) newMap.set(key, value);
    return new State(this._variables, newMap);
  }

  walk(variable) {
    if (this._values.has(variable))
      return this.walk(this._values.get(variable));
    else if (variable instanceof Pair)
      return new Pair(this.walk(variable.left), this.walk(variable.right));
    else
      return variable;
  }

  unify(a, b) {
    [a, b] = [this.walk(a), this.walk(b)];

    if (a === b) {
      return this;
    } else if (a instanceof Variable) {
      return this.assign([a, b]);
    } else if (b instanceof Variable) {
      return this.assign([b, a]);
    } else if (a instanceof Pair && b instanceof Pair) {
      let state = this.unify(a.left, b.left);
      return state && state.unify(a.right, b.right);
    } else {
      return null;
    }
  }

  results(n = this._variables.length) {
    return this._variables.slice(0, n).map(v => this.walk(v));
  }

  result(n = 0) {
    return this.results()[n];
  }

  inspect() {
    let pairs = Array.from(this._values).map(pair => pair.map(inspect).join(' = ')),
        vars  = inspect(this._variables);

    return `<State ${ vars } [${ pairs.join(', ') }]>`;
  }
}

File: haskell/State.hs
module State where

import Data.List
import Identifier
import Term


data State a  = State (Values a)
type Values a = [(Identifier, Term a)]

instance (Show a) => Show (State a) where
  show (State values) = "[" ++ (intercalate ", " (map format values)) ++ "]"
    where
      format (id, term) = (show id) ++ " -> " ++ (show term)

newState :: Values a -> State a
newState list = State list

bind :: String -> State a -> (State a, Identifier)
bind name (State values) =
  (State (values ++ [pair]), id)
  where
    id   = Identifier (length values) name
    pair = (id, Void)

mbind :: [String] -> State a -> (State a, [Identifier])
mbind names state =
  foldl extend (state, []) names
  where
    extend (state, ids) name =
      let (s, id) = bind name state in (s, ids ++ [id])

assign :: Identifier -> Term a -> State a -> State a
assign id@(Identifier index _) value (State values) =
  State (head ++ [mark] ++ tail)
  where
    head = take index values
    mark = (id, value)
    tail = drop (index + 1) values

unify :: (Eq a) => Term a -> Term a -> State a -> Maybe (State a)
unify x y state =
  let wx = walk x state; wy = walk y state in _unify state wx wy
  where
    _unify s (Var id)     y            = Just $ assign id y s
    _unify s x            (Var id)     = Just $ assign id x s
    _unify s (Pair l1 r1) (Pair l2 r2) = unify l1 l2 s >>= unify r1 r2
    _unify s x y  | x == y             = Just s
                  | otherwise          = Nothing

walk :: (Eq a) => Term a -> State a -> Term a
walk Void              _      = Void
walk None              _      = None
walk (Value v)         _      = Value v
walk (Pair left right) state  = Pair (walk left state) (walk right state)
walk var@(Var (Identifier index _)) state@(State values)
  | index < (length values)   = deref values index
  | otherwise                 = var
  where
    deref values index =
      let (id, val) = values !! index in
        if val == Void then Var id else walk val state

result :: (Eq a) => Int -> State a -> Term a
result index state = (results 0 state) !! index

results :: (Eq a) => Int -> State a -> [Term a]
results n s@(State values)
  | n == 0    = results (length values) s
  | otherwise = take n $ map _walk values
                where
                  _walk (var, val) = walk val s
